frontend/src/liboperator/operator/operator.go

```diff
@@ -480,6 +484,111 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
 	proxy.ServeHTTP(w, r)
 }
 
+// WebSocket endpoint that proxies ADB
+func adbProxy(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
+	vars := mux.Vars(r)
+	devId := vars["deviceId"]
+	dev := pool.GetDevice(devId)
+
+	if dev == nil {
+		http.Error(w, "Device not found", http.StatusNotFound)
+		return
+	}
+
+	devInfo := dev.privateData.(map[string]interface{})
+
+	// Find adb port for the device.
+	//   1. If registered device info has adb_port, use it.
+	//   2. If not, get index of the device and 6520 + index would be the adb port.
+	adbPort := 0
+	if adb_port, ok := devInfo["adb_port"]; ok {
+		adbPort = int(adb_port.(float64))
+	} else {
+		for i, id := range pool.DeviceIds() {
+			if id == devId {
+				adbPort = 6520 + i
+				break
+			}
+		}
+	}
+	if adbPort == 0 {
+		http.Error(w, "Cannot find adb port for the device", http.StatusNotFound)
+		return
+	}
+
+	// Prepare WebSocket and TCP socket for ADB
+	upgrader := websocket.Upgrader{}
+	wsConn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Print("Error while upgrading to WebSocket: ", err)
+		return
+	}
+	defer wsConn.Close()
+
+	tcpConn, err := net.Dial("tcp", fmt.Sprintf("127.0.0.1:%v", adbPort))
+	if err != nil {
+		log.Print("Error while connect to ADB: ", err)
+		return
+	}
+	defer tcpConn.Close()
+
+	// Redirect WebSocket to ADB tcp socket
+	redirectWsToTcp(wsConn, tcpConn)
+}
+
+func redirectWsToTcp(wsConn *websocket.Conn, tcpConn net.Conn) {
```

Comment:
Couldn't bi-directional data transmission be happened with `io.copy`? Just out of curiosity, I think it would make code much simpler if it's available.

Reply:
Changed to use io.Copy.